producer/servicelog/service_log_producer.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright © 2021 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package servicelog contains functions that can be used to produce (that is
+// send) messages to Service Log.
+package servicelog
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/RedHatInsights/ccx-notification-service/conf"
+	"github.com/RedHatInsights/ccx-notification-service/types"
+	httputils "github.com/RedHatInsights/insights-operator-utils/http"
+	"github.com/rs/zerolog/log"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// Producer is an implementation of Producer interface for Service Log
+type Producer struct {
+	Configuration           conf.ServiceLogConfiguration
+	AccessToken             string
+	TokenRefreshmentCounter int
+}
+
+// New constructs a new instance of Producer implementation
+func New(config conf.ConfigStruct) (*Producer, error) {
+	prod := &Producer{
+		Configuration: conf.GetServiceLogConfiguration(config),
+	}
+	err := prod.refreshToken()
+	if err != nil {
+		return prod, err
+	}
+	return prod, nil
+}
+
+// ProduceMessage sends the given message to Service Log
+func (producer *Producer) ProduceMessage(msg types.ProducerMessage) (partitionID int32, offset int64, err error) {
+	serviceLogURL := httputils.SetHTTPPrefix(producer.Configuration.URL)
+
+	client := &http.Client{
+		Timeout: time.Second * producer.Configuration.Timeout,
+	}
+
+	req, err := http.NewRequest(http.MethodPost, serviceLogURL, bytes.NewBuffer(msg))
+	req.Header.Add("Authorization", "Bearer "+producer.AccessToken)
+	if err != nil {
+		log.Error().Err(err).Msg("Got error while setting up HTTP request")
+		return -1, -1, err
+	}
+
+	response, err := client.Do(req)
+	if err != nil {
+		log.Error().Err(err).Msgf("Got error while making the HTTP request")
+		return -1, -1, err
+	}
+
+	switch response.StatusCode {
+	case http.StatusUnauthorized:
+		err = producer.refreshToken()
+		if err != nil {
+			log.Error().Msg("Access token could not be refreshed")
```

Comment:
maybe it's worth adding `.Err(err)` here?

Reply:
Yep, added :)